### PR TITLE
Add an SSL_SESSION accessor for obtaining the protocol version number…

### PR DESCRIPTION
**Pull request by Castaglia**
…, with

accompanying documentation.

### CODE REVIEW

**Comment by vdukhovni** [comment]
Should this function or a similar function also be able to indicate whether the session is a TLS or DTLS session? Though I guess applications could look at bit 15 (LSB==bit 0) to distinguish the two, in which case documentation to that effect might suffice.

File: doc/ssl/SSL_SESSION_get_protocol_version.pod [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+ #include <openssl/ssl.h>
+
+ int SSL_SESSION_get_protocol_version(const SSL_SESSION *s);
+
+=head1 DESCRIPTION
+
+SSL_SESSION_get_protocol_version() returns the time at which the session B<s> was
+established. The time is given in seconds since the Epoch and therefore
+compatible to the time delivered by the time() call.
```
Comment (mattcaswell) [inline doc/ssl/SSL_SESSION_get_protocol_version.pod]: Err...copy & paste error?

File: doc/ssl/SSL_SESSION_get_protocol_version.pod [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+the chosen value B<tm>.
+
+SSL_SESSION_get_timeout() returns the timeout value set for session B<s>
+in seconds.
+
+SSL_SESSION_set_timeout() sets the timeout value for session B<s> in seconds
+to B<tm>.
+
+The SSL_get_time(), SSL_set_time(), SSL_get_timeout(), and SSL_set_timeout()
+functions are synonyms for the SSL_SESSION_*() counterparts.
```
Comment (mattcaswell) [inline doc/ssl/SSL_SESSION_get_protocol_version.pod]: More copy & paste

File: doc/ssl/SSL_SESSION_get_protocol_version.pod [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+If the function is passed the NULL pointer for the session B<s>, 0 is returned.
+
+=head1 SEE ALSO
+
+L<ssl(3)>,
+L<SSL_CTX_set_timeout(3)>,
+L<SSL_get_default_timeout(3)>
+
+=head1 COPYRIGHT
```
Comment (mattcaswell) [inline doc/ssl/SSL_SESSION_get_protocol_version.pod]: Please add a HISTORY section stating that this function was added in 1.1.0

File: ssl/ssl_sess.c [hunk 4]
```diff
@@ -828,6 +828,13 @@ long SSL_SESSION_set_time(SSL_SESSION *s, long t)
     return (t);
 }
 
+int SSL_SESSION_get_protocol_version(const SSL_SESSION *s)
+{
+    if (s == NULL)
+        return (0);
```
Comment (mattcaswell) [inline ssl/ssl_sess.c]: Is this necessary. For other similar calls we just assume s is non-NULL, and it is a programmer error if a NULL is provided.

File: doc/ssl/SSL_SESSION_get_protocol_version.pod (hunk 3 above)
Comment (Castaglia) [inline doc/ssl/SSL_SESSION_get_protocol_version.pod]: Done.

File: doc/ssl/SSL_SESSION_get_protocol_version.pod (hunk 2 above)
Comment (Castaglia) [inline doc/ssl/SSL_SESSION_get_protocol_version.pod]: Oops!  Should be fixed now.  Thanks!

File: doc/ssl/SSL_SESSION_get_protocol_version.pod (hunk 1 above)
Comment (Castaglia) [inline doc/ssl/SSL_SESSION_get_protocol_version.pod]: Indeed.  Should be fixed now.

File: ssl/ssl_sess.c (hunk 4 above)
Comment (Castaglia) [inline ssl/ssl_sess.c]: The `SSL_SESSION_get_time` _et al_ functions do the `NULL` check; I was following that pattern.

Comment (mattcaswell) [inline ssl/ssl_sess.c]: Yes, that is inconsistent. As a general rule though we are trying to standardise on not doing the NULL check for the main object of a function, so I think we should remove this. There are other SSL_SESSION functions that don't do this.

Comment (richsalz) [inline ssl/ssl_sess.c]: Strongly agree with Matt :)

Comment (Castaglia) [inline ssl/ssl_sess.c]: Fair enough.  Code (and docs) updated appropriately.

**Comment by mattcaswell** [comment]
+1

**Comment by richsalz** [comment]
+1  nice doc!  merging.

**Pull request closed without merging**